src/qgraphicsmozview_p.cpp

```diff
@@ -937,27 +927,61 @@ void QGraphicsMozViewPrivate::touchEvent(QTouchEvent* event)
         }
     }
 
-    if (meventStart.mTouches.Length()) {
-        // We should append previous touches to start event in order
-        // to make Gecko recognize it as new added touches to existing session
-        // and not evict it here http://hg.mozilla.org/mozilla-central/annotate/1d9c510b3742/layout/base/nsPresShell.cpp#l6135
-        if (meventMove.mTouches.Length()) {
-            meventStart.mTouches.AppendElements(meventMove.mTouches);
+    // We should append previous touches to start event in order
+    // to make Gecko recognize it as new added touches to existing session
+    // and not evict it here http://hg.mozilla.org/mozilla-central/annotate/1d9c510b3742/layout/base/nsPresShell.cpp#l6135
+    QList<int> startIds(moveIds);
+
+    // Produce separate event for every pressed touch points
+    Q_FOREACH (int id, pressedIds) {
+        MultiTouchInput meventStart(MultiTouchInput::MULTITOUCH_START, timeStamp, TimeStamp(), 0);
+        startIds.append(id);
+        std::sort(startIds.begin(), startIds.end(), cmpInt);
+        Q_FOREACH (int startId, startIds) {
+            const QTouchEvent::TouchPoint& pt = event->touchPoints().at(idHash.value(startId));
+            mozilla::ScreenIntPoint nspt(pt.pos().x(), pt.pos().y());
+            meventStart.mTouches.AppendElement(SingleTouchData(pt.id(),
+                                                               nspt,
+                                                               mozilla::ScreenSize(1, 1),
+                                                               180.0f,
+                                                               pt.pressure()));
         }
-        Q_ASSERT(meventStart.mTouches.Length() > 0);
+
         ReceiveInputEvent(meventStart);
     }
-    if (meventMove.mTouches.Length()) {
-        if (meventStart.mTouches.Length()) {
-            meventMove.mTouches.AppendElements(meventStart.mTouches);
+
+    Q_FOREACH (int id, endIds) {
+        const QTouchEvent::TouchPoint& pt = event->touchPoints().at(idHash.value(id));
+        mozilla::ScreenIntPoint nspt(pt.pos().x(), pt.pos().y());
```

Comment:
There's no practical reason for that since touchend events contain only removed touch points whereas touchstart and touchmove contain all active touchpoints => no way to identify reliably touch points by their order.